### PR TITLE
fix/gas-esdt-asyncCallback

### PR DESCRIPTION
**Pull request by sasurobert**
added fix for gasRemaining computation in case of async callback and esdt transfer

### CODE REVIEW

File: integrationTests/multiShard/smartContract/polynetworkbridge/bridge_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	fmt.Printf("Initial minted sum: %s\n", initialVal.String())
+	integrationTests.MintAllNodes(nodes, initialVal)
+
+	round := uint64(0)
+	nonce := uint64(0)
+	round = integrationTests.IncrementAndPrintRound(round)
+	nonce++
+
+	// tokenManagerPath := "../testdata/polynetworkbridge/esdt_token_manager.wasm"
+	tokenManagerPath := "/home/camil.bancioiu/Work/Elrond/elrond-go/integrationTests/multiShard/smartContract/testdata/polynetworkbridge/esdt_token_manager.wasm"
```
Comment (dorin-iancu) [inline integrationTests/multiShard/smartContract/polynetworkbridge/bridge_test.go:62]: Use relative path?

**Review by dorin-iancu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

File: process/smartContract/process.go [hunk 2]
```diff
@@ -2216,6 +2229,9 @@ func (sc *scProcessor) EpochConfirmed(epoch uint32) {
 
 	sc.flagPenalizedTooMuchGas.Toggle(epoch >= sc.penalizedTooMuchGasEnableEpoch)
 	log.Debug("scProcessor: penalized too much gas", "enabled", sc.flagPenalizedTooMuchGas.IsSet())
+
+	sc.flagRepairCallBackData.Toggle(epoch >= sc.repairCallBackEnableEpoch)
+	log.Debug("scProcessor: repair call back", "enabled", sc.flagPenalizedTooMuchGas.IsSet())
```
Comment (SebastianMarian) [inline process/smartContract/process.go:2234]: flagRepairCallBackData instead flagPenalizedTooMuchGas

File: cmd/node/factory/structs.go [hunk 3]
```diff
@@ -1656,6 +1656,7 @@ func newShardBlockProcessor(
 		DeployEnableEpoch:              config.GeneralSettings.SCDeployEnableEpoch,
 		BuiltinEnableEpoch:             config.GeneralSettings.BuiltInFunctionsEnableEpoch,
 		PenalizedTooMuchGasEnableEpoch: config.GeneralSettings.PenalizedTooMuchGasEnableEpoch,
+		RepairCallbackEnableEpoch:      config.GeneralSettings.RepairAsyncCallBackEnableEpoch,
```
Comment (SebastianMarian) [inline cmd/node/factory/structs.go:1659]: config.GeneralSettings.RepairCallBackEnableEpoch instead config.GeneralSettings.RepairAsyncCallBackEnableEpoch

**Review by iulianpascalau — APPROVED** [review]
SC processor looks good, no backwards compatibility issues found

**Review by camilbancioiu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: process/smartContract/process.go [hunk 4]
```diff
@@ -807,6 +807,8 @@ func (sc *scProcessor) ExecuteBuiltInFunction(
 			return vmcommon.ExecutionFailed, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte("gas consumed exceeded"), snapshot, vmInput.GasLocked)
 		}
 
+		sc.penalizeUserIfNeeded(tx, txHash, newVMInput.CallType, newVMInput.GasProvided, newVMOutput)
```
Comment (iulianpascalau) [inline process/smartContract/process.go:810]: Builtin functions are already activated. Is this backwards compatibile?

File: process/smartContract/process.go [hunk 5]
```diff
@@ -1032,7 +1034,7 @@ func (sc *scProcessor) createVMInputWithAsyncCallBack(vmInput *vmcommon.Contract
 			CurrentTxHash:  vmInput.CurrentTxHash,
 		},
 		RecipientAddr:     vmInput.RecipientAddr,
-		Function:          "callback",
+		Function:          "callBack",
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1039]: backwards compatibility

File: process/smartContract/process.go [hunk 6]
```diff
@@ -823,6 +825,8 @@ func (sc *scProcessor) ExecuteBuiltInFunction(
 
 	isSCCallCrossShard := !isSCCallSelfShard && txTypeOnDst == process.SCInvoking
 	if !isSCCallCrossShard {
+		sc.penalizeUserIfNeeded(tx, txHash, newVMInput.CallType, newVMInput.GasProvided, newVMOutput)
```
Comment (iulianpascalau) [inline process/smartContract/process.go:828]: Also here. Shouldn't we used a flag?

File: process/smartContract/process.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/process.go]: done

File: process/smartContract/process.go (hunk 5 above)
Comment (sasurobert) [reply process/smartContract/process.go:1043]: this is ok. it happens only on ESDT

File: process/smartContract/process.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/process.go]: done

File: process/smartContract/process.go (hunk 5 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:1043]: ok

**Review by iulianpascalau — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by mateigraura — APPROVED** [review]
system test passed

**Pull request merged**